esp-hal-embassy/src/executor/thread.rs

```diff
@@ -140,7 +141,7 @@ This will use software-interrupt 3 which isn't available for anything else to wa
             unsafe { core::arch::asm!("waiti 0") };
         }
         // If this races and some waker sets the signal, we'll reset it, but still poll.
-        SIGNAL_WORK_THREAD_MODE[cpu].store(false, Ordering::SeqCst);
+        SIGNAL_WORK_THREAD_MODE[cpu].store(false, Ordering::Relaxed);
```

Comment:
The use of `asm!` might mean what I'm saying here is wrong but...

By using `Relaxed` here, you've given the compiler freedom to move this `store` operation *into* the if statement like so.

```
if load(Acquire) {
    store(false, Relaxed);
    wsr.ps
} else {
    store(false, Relaxed);
    waiti 0
}
```

And this can undo your change to reduce polling.

You can mark this as `Release` to make sure the if statement happened before this store.
Or add a comment saying that the `Relaxed` is still worth using.
Or say I'm wrong haha

Reply:
asm! makes the compiler assume all memory is touched and it won't reorder atomic accesses before it. Release has a surprisingly bad perf effect here.